bin/bl-exit

```diff
@@ -1,154 +1,425 @@
+            self.add_button(1, self.cancel_action, stock=gtk.STOCK_CANCEL)
+            self.build_button_visibility_array()
+            for button in self.bva:
+                (action, label, actionfunc, method, show, onError) = button
+                if not show == 0:
+                    self.add_button(show, actionfunc, label=label)
+            self.status = gtk.Label()
+            label_box = gtk.HBox()
+            label_box.pack_start(self.status)
+
+            vbox = gtk.VBox()
+            vbox.pack_start(self.button_box)
+            vbox.pack_start(label_box)
+
+            self.add(vbox)
+            self.show_all()
+
+        def build_button_visibility_array(self):
+            """Determine button visibily using bl-exit configuration file.
+            Build self.bva, an array of tuples, one entry per button,
+            containing (action, label, actionfunction, actionMethod, show,
+            onerror)
+            """
+            self.bva = []
+            bva = [('logout', '_Log out', self.logout_action),
+                   ('suspend', '_Suspend', self.suspend_action),
+                   ('hibernate', 'H_ibernate', self.hibernate_action),
+                   ('hybridsleep', 'H_ybrid sleep', self.hybridsleep_action),
+                   ('reboot', 'Re_boot', self.reboot_action),
+                   ('poweroff', '_Power off', self.shutdown_action)]
+            show_values = dict(never=0, always=1, maybe=2)
+            """Values that the 'show' keyword can take in the configuration
+            file."""
+            onerror_values = dict(novisual=0, visual=1)
+            """Values that the 'onerror' keyword can take in the configuration
+            file."""
+            # Per button default settings
+            per_button_defaults = dict(
+                logout='always',
+                suspend='always',
+                hibernate='never',
+                hybridsleep='never',
+                reboot='always',
+                poweroff='always'
+            )
+            for (action, label, actionfunction) in bva:
+                # Defaults.
+                show = show_values[per_button_defaults[action]]
+                onError = onerror_values['novisual']
+                if self.config:
+                    for section in ['default', action]:
+                        try:
+                            getshow = self.config.get(section, 'show')
+                            if getshow in show_values:
+                                show = show_values[getshow]
+                                if show == 2:
+                                    candoit = self.can_do_action(
+                                        actionToMethod[action])
+                                    if not candoit == 'yes':
+                                        show = 3
+                            else:
+                                pass
+                        except ConfigParser.NoSectionError as e:
```

Comment:
Maybe just catch a general parser error and bail out anyway. Saves one except clause.